q2_fmt/_peds.py

```diff
@@ -466,11 +479,458 @@ def _create_masking(time_metadata, donor_df, recip_df, reference_column):
+
+    Returns
+    -------
+    donor_mask: ndarray
+        a numpy array of donor feature information with the same order as
+        duplicated_recip_table. This allows for easy np array math.
+
+    Examples
+    --------
+    mismatched_df = pd.DataFrame({
+            'id': ["sample1", "sample1", "sample2", "sample2",
+                   "sample3", "sample3"],
+            "Ref": ["donor2", "donor3", "donor1", "donor3",
+                    "donor1", "donor2"]}
+                                ).set_index('id')
+    donor_df = pd.DataFrame({
+            'id': ['donor1', 'donor2', 'donor3'],
+            'Feature1': [1, 0, 0],
+            'Feature2': [0, 1, 0],
+            'Feature3': [0, 0, 1]}).set_index('id')
+
+    _create_sim_masking(mismatched_df, donor_df, reference_column)
+
+    ndarray[[0, 1, 0],
+            [0, 0, 1],
+            [1, 0, 0],
+            [0, 0, 1],
+            [1, 0, 0],
+            [0, 1, 0]]
+    """
+    donors = mismatched_df[reference_column]
+    donor_index_masking = []
+    for donor in donors:
+        donor_index_masking.append(donor_df.index.get_loc(donor))
+    donor_df_np = donor_df.to_numpy()
+    donor_mask = donor_df_np[donor_index_masking]
+    return donor_mask
+
+
+def _simulate_uniform_distro(mismatched_peds, iterations):
+    """Randomly samples the mismatched PEDS values.
+
+    Creates a uniform distribution of mismatched PEDS values by randomly
+    sampling N (iterations) times
+
+    Parameters
+    ----------
+    mismatched_peds: pd.DataFrame
+        A Dataframe that contains all mismatched PEDS values.
+    iterations: int
+        Number of iterations to run simulations (Number of times to randomly
+        sample mismatched_peds)
+
+    Returns
+    -------
+    peds_iters: pd.Series
+        a pd.Series with all iterations of mismatched PEDS Values. This will
+        later be compared to an actual PEDS value.
+
+    Examples
+    --------
+    mismatched_peds = [0, 0, 0,0 ]
+    iterations = 10
+
+    _simulate_uniform_distro(mismatched_peds, iterations)
+
+    pd.Series([data = [0, 0, 0, 0, 0, 0, 0, 0, 0],
+               index = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]])
+    """
+    peds_iters = random.choices(mismatched_peds, k=iterations)
+    # Tranforming to series for easy series math in _per_subject_stats()
+    peds_iters = pd.Series(peds_iters)
+    return peds_iters
+
+
+def _per_subject_stats(mismatched_peds, actual_temp,
+                       iterations, mismatched_pairs_n):
+    """Creates per subject PEDS stats
+
+    Creates per subject PEDS stats by sampling mismatch PEDS values 'iteration'
+    number of time. The list of mismatched PEDS values is compared to the
+    actual PEDS value. The Test Statistic for this is the iterations - the
+    number of mismatched PEDS value that are greater than the actual p-value.
+    p-values are corrected by adding 1 to each side of the faction
+    because Monte Carlo Simulation is able to produce an invalid p-value of 0.
```

Comment:
I would add the formula to the doc string to make it easier to follow what's happening.